src/webgpu/util/compare.ts

```diff
@@ -132,7 +132,13 @@ export function compare(got: Value, expected: Value, cmpFloats: FloatMatch): Com
 export function anyOf(...expectations: (Value | Comparator)[]): Comparator {
   return (got, cmpFloats) => {
     const failed: Array<string> = [];
+    const s = new Set<String>();
     for (const e of expectations) {
+      if (s.has(e.toString())) {
```

Comment:
I am a touch leery of using string matching to do filtering on numbers, since I am not sure we guarantee that Value.toString() is unique for every possible value, since displayed digits might being truncated. If you could confirm that we are not truncating information that would alay my concern.

Though as I think about it, I think there is a bigger issue. Comparators don't define a .toString(), so I think you will get undefined for any e that is a Comparator, so you might be skipping later Comparators. I would move the check into the e as Value branch below to avoid this.

Reply:
> I am a touch leery of using string matching to do filtering on numbers, since I am not sure we guarantee that Value.toString() is unique for every possible value, since displayed digits might being truncated. If you could confirm that we are not truncating information that would alay my concern.

We defined toString() for Scalar which outputs (among other things) the 64-bit hex value. We also define toString for Vector in terms of the Scalar one. So every value will have a unique string.

> Though as I think about it, I think there is a bigger issue. Comparators don't define a .toString(), so I think you will get undefined for any e that is a Comparator, so you might be skipping later Comparators. I would move the check into the e as Value branch below to avoid this.

Comparators are functions, and their toString() actually outputs the definition of the function itself. I tested this locally by doing this in makeCase:

```
    const makeCase = (y: number, x: number): Case => {
      const c = makeBinaryF32Case(y, x, Math.atan2, true);
      const c2 = makeBinaryF32Case(y, x, Math.atan2, true);
      c.expected = anyOf(c.expected, c2.expected);
      return c;
    };
```

When it failed, it only output one value. However, I will admit that it's a bit confusing with the comparator in there. How about this instead: instead of skipping based on the `toString()` of `expectations` (which includes comparators), we skip the adding of `cmp.expected` to the `failed` array. This way, we're already dealing with strings, and just avoiding adding duplicate strings to the failed list. I'll do that, it's simpler and easier to understand.

Reply:
Updated with a new version that just avoids adding the same expected values to `failed`, which is now a `Set<string>` instead of an `Array<string>`. Note that the order of expected values should be in order of insertion, according to the [documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).